command/agent.go

```diff
@@ -805,25 +775,47 @@ func (c *AgentCommand) Run(args []string) int {
 		close(c.startedCh)
 	}
 
-	// Listen for signals
-	// TODO: implement support for SIGHUP reloading of configuration
-	// signal.Notify(c.signalCh)
-
 	var g run.Group
 
+	g.Add(func() error {
+		for {
+			select {
+			case <-c.SighupCh:
+				c.UI.Output("==> Vault Agent config reload triggered")
+				err := c.reloadConfig(c.flagConfigs)
+				if err != nil {
+					c.outputErrors(err)
+				}
+				// Send the 'reloaded' message on the relevant channel
+				select {
+				case c.reloadedCh <- struct{}{}:
+				default:
+				}
+			case <-ctx.Done():
+				c.notifySystemd(systemd.SdNotifyStopping)
+				return nil
+			}
+		}
+	}, func(error) {
+		cancelFunc()
+	})
+
 	// This run group watches for signal termination
 	g.Add(func() error {
 		for {
 			select {
 			case <-c.ShutdownCh:
-				c.UI.Output("==> Vault agent shutdown triggered")
+				c.UI.Output("==> Vault Agent shutdown triggered")
 				// Notify systemd that the server is shutting down
 				c.notifySystemd(systemd.SdNotifyStopping)
 				// Let the lease cache know this is a shutdown; no need to evict
 				// everything
 				if leaseCache != nil {
 					leaseCache.SetShuttingDown(true)
 				}
+				if c.reloadedCh != nil {
```

Comment:
Is this necessary?  I'd probably remove it since reloadedCh is only needed for tests, and I assume they don't depend on this behaviour.

Reply:
No it can be removed, I just added it as a bit of a tidy up during shutdown - but you're right. :D

Reply:
It's partly that this impeded readability by virtue of adding unnecessary complexity.  It's also partly that when I see code like this, I automatically wonder: when do we assign nil to the channel?  If we don't assign nil to it, can we panic due to a double close?  Is there a lock protecting the assignment? etc